### PR TITLE
CHE-5810: Improve client encoding for file or folder path to support special symbols

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
Improve client encoding for file or folder path to support special symbols:  "%", "[", "]", "#", ";".

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5810

#### Changelog
Improve client encoding for file or folder path to support special symbols:  "%", "[", "]", "#", ";".

#### Release Notes
Improve client encoding for file or folder path to support special symbols:  "%", "[", "]", "#", ";".

#### Docs PR
Don't need bug Fix.

### CODE REVIEW

**Review by azatsarynnyy — APPROVED** [review]

File: ide/commons-gwt/src/main/java/org/eclipse/che/ide/resource/Path.java [hunk 1]
```diff
@@ -1143,4 +1176,24 @@ public static Path commonPath(Path... paths) {
 
     return commonPath;
   }
+
+  /** Returns path encoded by segments without device. */
+  public String getEncodedPath() {
```
Comment (vzhukovs) [inline ide/commons-gwt/src/main/java/org/eclipse/che/ide/resource/Path.java]: Is it tested?

Create separate utility class called `PathEncoder` and move this method as static there. See how `java.net.URLEncoder#encode(java.lang.String, java.lang.String)` method done.

This is not a right place for this method, it should be a static utility method.

**Review by vzhukovs — CHANGES_REQUESTED** [review]

File: ide/commons-gwt/src/main/java/org/eclipse/che/ide/resource/Path.java (hunk 1 above)
Comment (AndrienkoAleksandr) [reply ide/commons-gwt/src/main/java/org/eclipse/che/ide/resource/Path.java]: Yes it was tested.

**Review by tolusha — APPROVED** [review]

**Review by vparfonov — APPROVED** [review]

File: ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import static com.google.gwt.http.client.URL.encodePathSegment;
+import static org.eclipse.che.ide.resource.Path.SEPARATOR;
+
+import org.eclipse.che.ide.resource.Path;
+
+/**
+ * @author Alexander Andrienko
+ * @author Mykola Morhun
+ */
+public class PathEncoder {
```
Comment (tolusha) [inline ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java:22]: Add several tests also.

**Review by vzhukovs — APPROVED** [review]

File: ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java (hunk 2 above)
Comment (mmorhun) [reply ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java:22]: Well, we have only one method to test and it uses native javascript method. I think it won't worst to try to deal with it in tests.

Comment (mmorhun) [reply ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java:22]: But I tried...

Comment (tolusha) [reply ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java:22]: It is pure java code. Why do you need js?

Comment (mmorhun) [reply ide/commons-gwt/src/main/java/org/eclipse/che/ide/util/PathEncoder.java:22]: Look deeper. `encodeComponent` method uses native js method under the hood.

**Pull request merged**